test/org/sagebionetworks/bridge/services/ExternalIdServiceTest.java

```diff
@@ -63,13 +71,137 @@ public void before() {
+        
+        // verify
+        assertEquals(SUBSTUDY_ID, extId.getSubstudyId());
+        assertEquals(1, account.getAccountSubstudies().size());
+        
+        AccountSubstudy acctSubstudy = Iterables.getFirst(account.getAccountSubstudies(), null);
+        assertEquals(TestConstants.TEST_STUDY_IDENTIFIER, acctSubstudy.getStudyId());
+        assertEquals(ID, acctSubstudy.getExternalId());
+        assertEquals(SUBSTUDY_ID, acctSubstudy.getSubstudyId());
+        
+        verify(externalIdDao).createExternalId(extId);
+        verify(accountDao).updateAccount(account, null);
+    }
+    
+    @Test
+    public void migrateExternalIdentifierWithExistingSubstudyAssociation() throws Exception {
+        // setup
+        when(substudyService.getSubstudy(TestConstants.TEST_STUDY, ID, true))
+            .thenReturn(Substudy.create());
+        
+        ExternalIdentifier extId = ExternalIdentifier.create(TestConstants.TEST_STUDY, ID);
+        when(externalIdDao.getExternalId(TestConstants.TEST_STUDY, ID))
+            .thenReturn(extId);
+        
+        Account account = Account.create();
+        account.setId(USER_ID);
+        // This exists, but has no external ID
+        AccountSubstudy acctSubstudy = AccountSubstudy.create(TestConstants.TEST_STUDY_IDENTIFIER, SUBSTUDY_ID, USER_ID);
+        account.getAccountSubstudies().add(acctSubstudy);
+        
+        AccountId accountId = AccountId.forExternalId(TestConstants.TEST_STUDY_IDENTIFIER, ID);
+        when(accountDao.getAccount(accountId)).thenReturn(account);
+        
+        // execute
+        externalIdService.migrateExternalIdentifier(study, ID, SUBSTUDY_ID);
+        
+        // verify
+        assertEquals(SUBSTUDY_ID, extId.getSubstudyId());
+        assertEquals(1, account.getAccountSubstudies().size());
+        assertEquals(ID, account.getExternalId());
+        
+        acctSubstudy = Iterables.getFirst(account.getAccountSubstudies(), null);
+        assertEquals(TestConstants.TEST_STUDY_IDENTIFIER, acctSubstudy.getStudyId());
+        assertEquals(ID, acctSubstudy.getExternalId()); // now it does
+        assertEquals(SUBSTUDY_ID, acctSubstudy.getSubstudyId());
+        
+        verify(externalIdDao).createExternalId(extId);
+        verify(accountDao).updateAccount(account, null);
+    }
+    
+    @Test
+    public void migrateExternalIdentifierFromSingularToSubstudyAssociation() throws Exception {
+        // setup
+        when(substudyService.getSubstudy(TestConstants.TEST_STUDY, ID, true))
+            .thenReturn(Substudy.create());
+        
+        ExternalIdentifier extId = ExternalIdentifier.create(TestConstants.TEST_STUDY, ID);
+        when(externalIdDao.getExternalId(TestConstants.TEST_STUDY, ID))
+            .thenReturn(extId);
+        
+        Account account = Account.create();
+        account.setId(USER_ID);
+        // This was already a managed external ID, and will be updated correctly
+        account.setExternalId(ID);  
```

Comment:
Would it be better if this starts as a different external ID so we can verify it got replaced?

Reply:
Changed this along with a comment about how this would be remarkable in production (it would be ok, but we aren't exporting multiple external IDs yet, so if this method found an account under ID #1, and yet it had ID #2 in the account.externalId field, this call is probably fixing something that's broken).